language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,370 @@
+    vArg <- normalizeTypeM $ void arg
+    checkTypeM ann term =<< unfoldFixOf vPat vArg k
+    pure $ TyIFix () <$> vPat <*> vArg
+inferTypeM (Unwrap ann term)        = do
+    vTermTy <- inferTypeM term
+    case unNormalized vTermTy of
+        TyIFix _ vPat vArg -> do
+            k <- inferKindM $ ann <$ vArg
+            -- Subparts of a normalized type, so normalized.
+            unfoldFixOf (Normalized vPat) (Normalized vArg) k
+        _                  -> throwing _TypeError (TypeMismatch ann (void term) (TyIFix () dummyType dummyType) vTermTy)
+inferTypeM (Error ann ty)           = do
+    checkKindM ann ty $ Type ()
+    normalizeTypeM $ void ty
+-- | PORT ENDS HERE. Taken from `Language.PlutusCore.Typecheck.Internal`
+--------------------------------------------------------------------------
+
+-- Note on symbols:  '=>' means implies
+
+-- checkKindBinding(G,b)
+-- vB=normalizeBinding(G,b)  checkTypeBinding(G, vB)
+-- !null(bs) => [infer| G,withNewType(b), withNewVars(vB) !- (let nonrec {bs} in inT) : ty]
+-- null(bs) => [infer| G,withNewType(b), withNewVars(vB) !- inT : ty]
+-- ty ~>? vTy
+-- -------------------------------------------------
+-- [infer| G !- (let nonrec {b ; bs} in inT) : vTy]
+inferTypeM (Let _ NonRec bs inTerm) =
+    -- Check each binding individually, then if ok, introduce its new type/vars to the (linearly) next let or inTerm
+    foldr checksThenScope (inferTypeM inTerm) bs
+ where
+   checksThenScope :: Binding TyName Name uni ann -> TypeCheckM uni e res -> TypeCheckM uni e res
+   checksThenScope b acc = do
+       -- check the kind of the newly-introduced typevariable (if there is one) is correct
+       checkKindBinding b
+       -- normalize all subtypes of binding
+       normB <- normalizeBinding NonRec b
+       -- check the type of newly-introduced variable term (if there is one)
+       checkTypeBinding normB
+       -- Checks pass, create new scope and scope acc
+       newVars <- bindingNewVars normB
+       withNewType (void b) .  -- Note: the order of adding to env here does not matter, also b or normB does not matter
+           withVars newVars $
+               acc
+
+-- G'=G,withNewTypes(bs)
+-- forall b in bs. checkKindBinding(G', b)
+-- νBs=forall b in bs. normalizeBinding(G', b)
+-- G''=G',withNewVars(vBs)
+-- forall vB in vBs. checkTypeBinding(G'', vB)
+-- [infer| G'' !- inT : ty] ty ~>? vTy
+-- -------------------------------------------------
+-- [infer| G !- (let rec bs in inT) : vTy]
+inferTypeM (Let _ Rec bs inTerm) =
+    withNewTypes (void <$> bs) $ do
+       -- with all new types in scope (since we are rec), make sure these newly-introduced types are correctly kinded
+       for_ bs checkKindBinding
+       -- normalize all bindings
+       normBs <- for bs $ normalizeBinding Rec
+       -- collect all newly-introduced Variables, maybe rename their normalized types (for global-uniqueness)
+       newVars <- mconcat . toList <$> for normBs bindingNewVars
```

Comment:
Hmmmm good call. I did this because normalizeBinding relies on kinds (which in Rec case are in scope), and I wanted to make sure that all the kinds are correct **before** continuing with normalization. If you think nothing will go wrong, I can merge this *fors* into one. It just looked more safe to me.

Reply:
That actually makes a lot of sense. Please add this as an inline comment. I don't think we need to change anything here then as what I proposed sounds dangerous.

Reply:
> That actually makes a lot of sense. Please add this as an inline comment. I don't think we need to change anything here then as what I proposed sounds dangerous.

I wrote that comment to @bezirg before @michaelpj's comment pop up.

@michaelpj 

> But you can check that the kind of a given binding is correct before normalizing it, you don't have to check that the kinds of all the bindings are correct before normalizing the types. I think.

What if you define a type `ty` of kind `*` that uses a fixed-point combinator defined later in the mutual block? Then the fixed-point combinator is not type checked before you start normalizing `ty` and so you might loop.

Does that make sense?

Reply:
Ah yes, you're right. If it's recursive we do need to do all the kind checks first!